doc/design/dist/README.md

```diff
@@ -0,0 +1,190 @@
+- Periodically save its parameters to distributed file system by overriding the previous save.
+
+### Optimization Algorithms
+
+The communication pattern between the trainers and the parameter servers depends on the category of optimization algorithm:
+
+- Synchronous Stochastic Gradient Decent (sync-SGD)
+
+	Parameter server will wait for all trainer finish n-th mini-batch calculation and send their gradients before broadcasting new parameters to every trainer. Every trainer will wait for the new parameters before starting n+1-th mini-batch.
+  
+- Asynchronous Stochastic Gradient Decent (async-SGD)
+
+	There will no synchronization between different trainers, and parameter server updates its parameter as soon as it receives new gradient:
+
+	- Each trainer uploads its accumulated gradient every **n** mini-batches.
+	- Every **m** mini-batches, the trainer downloads new parameters from parameter server.
+	- **n** and **m** do not have to be equal.
+
+## Fault Tolerant
+
+The training job will pause if the master process is dead, or any of the parameter server process is dead. They will be started by the cluster management system and recover in few minutes. Please refer to [fault recovery](#fault-recovery).
+
+The training job will continue to make progress if there is at least one training process running. The strategy depends on the type of optimization algorithm:
+
+- sync-SGD
+
+	TODO
+
+- async-SGD
+
+	Since async-SGD does not require synchronization between mini-batches, the system will by definition make process if at least one trainer is running.
+
+## Fault Recovery
+
+PaddlePaddle uses [etcd](https://github.com/coreos/etcd) to keep track of the states of processes. Because etcd is a distributed reliable key-value store, the restarted process can recover its states from etcd. The model parameter are periodically saved into distributed file system, so a restarted parameter server can recover its parameters from the saved file.
+
+Now we will introduce how each process recovers from failure:
+
+<img src="src/paddle-etcd.png" width="650"/>
+
+### Master Process
+
+When the master is started by the cluster management system, it executes the following steps at startup:
+
+1. Grabs a unique *master* lock in etcd, which prevents concurrent master instantiations.
+1. Recovers the task queues from etcd if they already exists, otherwise the master will create them.
+1. Watches the trainer prefix keys `/trainer/` on etcd to find the live trainers.
+1. Starts dispatching the tasks to the trainers.
+
+The master process will kill itself if its etcd lease expires.
+
+When the master process is dead for any reason, the cluster management system will restart it. It will be online again with all states recovered from etcd in few minutes.
+
+### Trainer Process
+
+When the trainer is started by the cluster management system, it executes the following steps at startup:
```

Comment:
In my imagination, the user will start the trainers and the parameter servers. Underlying implementation could be from a k8s config xml created and launched from user's laptop. Not from the master. --- let's discuss this if you have a different idea. I am not very sure about the benefit of using the master to start the trainer and the parameter servers --- seems the master is trying to do to many things at once in this case, but I may be wrong.

The number of desired parameter servers can be an environment variable for the master process. And the master process can know the number of current parameter servers from watching etcd `/ps/` and `/ps_candidate/` prefix.
The master does not need to know what is the number of desired trainers. It can know the number of current trainers by watching etcd `/trainer/` prefix.

About setting up some parameter server process before user submission, I think we will need to pre-setup parameter server process if the setup time is high, so we can pre-setup to save some time. But the setup time for parameter server does not seem to be high?